metadata-io/src/main/java/com/linkedin/metadata/entity/DeleteReferencesService.java

```diff
@@ -0,0 +1,177 @@
+      return new Aspect((DataMap) newValue);
+    } catch (CloneNotSupportedException e) {
+      return new Aspect();
+    }
+  }
+
+  /**
+   * This method chooses which sub method to invoke based on the path being iterated on.
+   *
+   * @param value           The value to be removed from the data complex object. Merely propagated down in this method.
+   * @param schema          The schema of the data complex being visited. Merely propagated down in this method.
+   * @param o               The data complex instance with the actual data being visited.
+   * @param pathComponents  The list of sub-strings representing the path from the root of the data complex to the value
+   *                       to be removed.
+   * @param index           The current index in the list of path components the method is meant to process.
+   * @return  The updated DataComplex object without the value given as input.
+   */
+  private static DataComplex removeValueBasedOnPath(String value, DataSchema schema, DataComplex o, List<String> pathComponents,
+      int index) {
+
+    final String subPath = pathComponents.get(index);
+
+    // Processing an array
+    if (subPath.equals("*")) {
+      // Process each entry
+      return removeValueFromArray(value, (ArrayDataSchema) schema, (DataList) o, pathComponents, index);
+    } else { // Processing a map
+      return removeValueFromMap(value, (RecordDataSchema) schema, (DataMap) o, pathComponents, index);
+    }
+  }
+
+  /**
+   * This method is used to visit and remove values from DataMap instances if they are the leaf nodes of the original
+   * data complex object.
+   *
+   * Note that this method has side effects and mutates the provided DataMap instance.
+   *
+   * @param value           The value to be removed from the data map object.
+   * @param spec            The schema of the data complex being visited. Used to get information of the optionallity of
+   *                        the data map being processed.
+   * @param record          The data list instance with the actual data being visited.
+   * @param pathComponents  The list of sub-strings representing the path from the root of the data complex to the value
+   *                       to be removed.
+   * @param index           The current index in the list of path components the method is meant to process.
+   * @return  The updated DataComplex object without the value given as input.
+   */
+  private static DataComplex removeValueFromMap(String value, RecordDataSchema spec, DataMap record, List<String> pathComponents,
+      int index) {
+    // If in the last component of the path spec
+    if (index == pathComponents.size() - 1) {
+      // Can only remove if field is optional or if (parent is optional AND struct will be empty or just with optionals)
+      boolean canDelete = spec.getField(pathComponents.get(index)).getOptional();
```

Comment:
this boolean suggests that we can only delete if the field is optional - but the comment suggests otherwise. help me understand?